WIP - Convert to ES6 and JSX

mjameswh opened the pull request:
Converted code base from CoffeeScript to ES6 and JSX, and upgraded to React 16. Fix #484, #278 and #180.

This is still work in progress; however, given that it is a very intrusive change, I hope to solicit comments as soon as possible. Also, it seems rather stable and mostly functional to me, but the conversion process has been lengthly and I certainly introduced a few small bugs during the way; testing in multiple environment is therefore desirable.

At present, I consider that I have the following items to complete before this can be considered for merging to master:

- [x] Fix the option panel in React GUI (it currently doesn't show)
- [x] Remove superfluous `return` statements added almost everywhere by the decaffeination process
- [x] Fix remaining IIFE blocks (that is, for loops in synthetic functions) introduced by the decaffeination process
- [x] Fix export statements to avoid duplication
- [x] Fix usages of Array.includes (several were added by the decaffeination process, also see #503)
- [ ] Make the code generally look more like native JavaScript code
- [ ] Introduce Webpack and make it generates an all-in-one distribution file, to be used for demo pages
- [ ] Update Read Me file appropriately

mjameswh committed:
Preparation for decaffeination: rename all coffee files to .js

mjameswh committed:
Decaffeination

mjameswh committed:
ESLint

- Added eslint, with some opinionated but descent configuration
- Fixed all eslint issues that could be fixed automatically (mostly
indentation and semicolons)

mjameswh committed:
Fix undesirable artefacts resulting from decaffeination

- Removed all synthetic « Array.from(…) » and « || [] » patterns
- Replace the pattern
      let something;
      …;
      export default (something = function() {  // or something = class
…
          …
      })
  by either a class definition or a const assignment, first,
  then an « export default » on a distinct line.
- Remove all comments about DS102 (implicit returns) and DS207
  (shorter null checks) because these will be removed during a later
  code review; the comment simply won’t help in identifying fix
  opportunities
- Use (() => {}) for noop functions instead of « function() {} »
- Replaced most remaining « require(…) » by « import(…) »
   (left shims files and createFactory(require(…)) patterns untouched)
- Fix some indentation and empty lines
- Fix all switch-in-dummy-functions patterns (DS205 - IIFEs)
- Marked all push-in-dummy-functions patterns (DS205 - IIFEs)
   These will however be fixed in a subsequent pass.
- Removed some unused imports
- Replace all uses of « module.exports » by ES6 style exports
- Remove all constructor hacks (DS001) and add missing super()

mjameswh committed:
Upgraded to React 16

- Removed all React related shims

mjameswh committed:
Convert all react components to regular ES6 classes

- Replace all uses of createReactClass(…) by regular ES6 classes
extending React.Component.
- Add « UNSAFE_ » prefix on React functions that have been deprecated

Note: at this step, I used only component classes (aka stateful
components). At a later step, many will be converted again, this time
to functional style components (aka stateless components).

mjameswh committed:
Converted all React code to JSX

- Convert all React rendering to JSX
- Removed all references to DOM (react-dom-factories) and
React.createFactory
- Removed specialized classes previously created to handle dual
behaviour buttons (zoom in/out and undo/redo)
- Replace usage of PureRenderMixin by extending React.PureComponent
- Bound several callback functions in constructor (probably missed some)
- Other local refactoring that made the conversion easier

mjameswh committed:
Reviewed Gulp file to remove all Coffee related steps

mjameswh committed:
Abolished use of mixing in react classes

- Inlined the createSetStateOnEventMixin mixin in every classes that
   previously used it

mjameswh committed:
Fix Gulp's watch to also monitor jsx files

mjameswh committed:
Cleanup of project dependencies

- Remove some remaining references to libraries that have already been discarded
- Introduced browserify-shim to handle getting access to React/ReactDOM
  when these are loaded as a distinct script tag in the html file
- Remove all third-party libraries artefacts included for demo purpose;
  these are heavy, hard to maintain and simply useless; link to CDN instead.

mjameswh committed:
Temporarily fix import/export issues resulting

These will have to be reconsidered at a latter step; not sure why
duplication of exports are now required (that is, default vs non default;
why was it not required before?). (+1 squashed commit)
Squashed commits:
[0f0f13f] Temporarily fix import/export issues resulting

These will have to be reconsidered at a latter step; not sure why
duplication of exports are now required (that is, default vs non default;
why was it not required before?).

mjameswh committed:
Make React classes complies with their requirements regarding ES6

- Replace getInitialState() by setting the initial state in the constructor,
  and possibly have add getDerivedStateFromProps() if the initial state
  is based on the component's props. Similarly, removed usage of
  componentWillReceiveProps() which has been deprecated.
- Replace getDefaultProps() by setting the defaultProps static field on
  the component class.
- Use HOC for components that requires pre-configuration (ToolButton, Options...)
- Make sure callbacks are bound to this in the constructor
- Fix several cases where a value was set as a class field, rather than
  a default instance value (that is, on the class' prototype).

mjameswh committed:
Fix various minor bugs introduced in the conversion process

irskep commented:
![it's happening!](https://user-images.githubusercontent.com/40085/42062813-c3c38c44-7ae3-11e8-8787-5f587de88ec5.gif)

irskep commented on demo/core_demo.html:
```diff
@@ -124,8 +124,8 @@
-    <script src="/demo/jquery-1.8.2.js"></script>
-    <script src="/demo/fastclick.js"></script>
+    <script src="https://code.jquery.com/jquery-1.8.3.js" crossorigin></script>
+    <script src="https://unpkg.com/fastclick@1.0/lib/fastclick.js" crossorigin></script>
```
The reason these are included in the repo is so you can run the demo without an internet connection, which I've used many times on airplanes. So I'd rather not have this change included in this PR, as it would be a separate discussion.

mjameswh replied:
Fair enough... Then I will modify the gulp file to automatically generate generate those library files, based on the versions specified in the package file. Or we could even generate an all-in-one distribution file, including React and others... Do you need to be able to test be demo with multiple versions or having the latest version only is enough?

irskep replied:
Looks like my earlier reply got messed up by Github's email gateway. Latest version is fine!

irskep commented:
Great idea. Latest is fine!

On Thu, Jun 28, 2018, at 3:22 PM, James Watkins-Harvey wrote:
> *@mjameswh* commented on this pull request.


> 
> In demo/core_demo.html[1]:


> > @@ -124,8 +124,8 @@ style="display: inline-block; border: 1px solid
> > yellow"></div> </div>  -    <script src="/demo/jquery-
> > 1.8.2.js"></script> -    <script src="/demo/fastclick.js"></script>
> > +    <script src="https://code.jquery.com/jquery-1.8.3.js"
> > crossorigin></script> +    <script src="
> > https://unpkg.com/fastclick@1.0/lib/fastclick.js"
> > crossorigin></script>
>> Fair enough... Then I will modify the gulp file to automatically
> generate generate those library files, based on the versions specified
> in the package file. Or we could even generate an all-in-one
> distribution file, including React and others... Do you need to be
> able to test be demo with multiple versions or having the latest
> version only is enough?> — You are receiving this because you commented. Reply to this email
> directly, view it on GitHub[2], or mute the thread[3].

Links:

  1. https://github.com/literallycanvas/literallycanvas/pull/511#discussion_r199004508
  2. https://github.com/literallycanvas/literallycanvas/pull/511#discussion_r199004508
  3. https://github.com/notifications/unsubscribe-auth/AACclU7rC_FjkQicBxYiurBzfw4CIl4Nks5uBVcegaJpZM4U8Hbn

mjameswh committed:
Removed spurious returns added by the decaffeination process

Thanks but no thanks, CoffeeScript, for implicit returns...

mjameswh committed:
Fixed all IIFE blocks (for-in-synthetic-function)

Note: IIFE blocks are structures created by the decaffeination process in
order to deal with structures (switch/for/if...) being used as regular
expressions.

Also found and fixed some spurious return statements.

mjameswh committed:
Fix all usages of Array.includes (replaced by Array.indexOf()) - Fix #503

rohitharkhani commented:
Coll This fixing localisation bug also.

mjameswh committed:
Fix exports duplication

Also fix some issues introduced in previous commits:
  - polygons wont work, due to a typo ("foreach" instead of "forEach)
  - typo error in package.json

mjameswh committed:
Fix three escaping issues highlighted by eslint

mjameswh committed:
Fix an issue with Polygon tool introduced in a previous commit

mjameswh committed:
WIP - Converting from Gulp+Browserify to Webpack

mjameswh committed:
Ported PR #514 and #517

mjameswh committed:
Fix two minor issues that caused ESLint errors

mjameswh commented:
Just a note, I wont be able to work on this for the next few days, due to a medical issue.

@irskep If you prefer, feel free to apply this PR to a distinct branch (say... 0.6.x), or even, create a branch 0.5.x from master, and apply the PR directly to master. This will certainly encourage other people to use this code base for further contributions, and hopefully, some might even give help on getting release 0.6.0 ready ;)

ggomez91 commented:
> Just a note, I wont be able to work on this for the next few days, due to a medical issue.
> 
> @irskep If you prefer, feel free to apply this PR to a distinct branch (say... 0.6.x), or even, create a branch 0.5.x from master, and apply the PR directly to master. This will certainly encourage other people to use this code base for further contributions, and hopefully, some might even give help on getting release 0.6.0 ready ;)

👍 This would be cool

mjameswh commented:
Unfortunately, due to personal reasons, I have missed my opportunity window for working on this project. I have no idea now when I will be able to put more concrete efforts on this. Yet, both decaffeination and the build process review should be pretty complete.

I don't remember exactly, but I think the main outstanding issue is the fact that the old build process used a precompiler to determine either to build the core or the gui variant. I have not found how to adapt this process to the webpack build system, and I personally think that it would be preferable not to do so. It should be easy and much more desirable to adapt the project's structure so that core and gui are put in distinct subdirectories and gui is built atop the core library (as intended in #377).

Anyway, if @ggomez91 is willing to continue the effort, then please don't refrain. Also, if you have questions, I might still be able to assist.

jasoniangreen commented:
I'd love to see this merged. What's the biggest outstanding issues stopping that. Maybe I can help out.
If we leave webpack as is, is this the only thing to do?

> It should be easy and much more desirable to adapt the project's structure so that core and gui are put in distinct subdirectories and gui is built atop the core library (as intended in #377).

irskep commented:
I haven't checked the branch out, but I think it's correct; someone needs to tease apart the code into a "core" library, and a "gui" library, which have the right set of dependencies and APIs.

There are real-world users of the core build, but another option would be to bump the version number and temporarily remove support for it.

jasoniangreen commented:
Would you be against the core library being in a separate repo? That way we wouldn't have to figure out how to mix old and new and it could just pull it in as a dependency. You could still make that dependency available in the same path so you don't break backward compatibility.

jasoniangreen commented:
Actually I should play around with what's here first before making any suggestions.

irskep commented:
I would expect the core build to be moved to a new package in a separate repo.

jasoniangreen commented:
Ah perfect. I'll see if I can get familiar with the code and see how the dependencies map out.

jasoniangreen commented:
@irskep how about you create an empty repo in this org called literallycanvas-core so I have something to make a PR to.

irskep commented:
Good idea. I pushed a clone of the whole repo here, so history can be maintained: https://github.com/literallycanvas/literallycanvas-core

jasoniangreen commented:
I shouldn't need any css in the core, should I? And all the demo stuff, are you ok with that just being in the main repo?

irskep commented:
Yeah, `core` should be more of a "drawing view construction kit" than "already-usable drawing view GUI." No CSS, no images, no demos.
